contrib/resourcetuner/cgroups.go

```diff
@@ -0,0 +1,161 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:build linux
+
+package resourcetuner
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/containerd/cgroups/v3/cgroup2"
```

Comment:
Naive general cgroup question from someone that doesn't know much about cgroups (made as PR thread on this line because it's easier than top-level PR comment), does querying system limits inside a cgroup not represent the cgroup limit? What is the difference in a container between the observable system stats and the cgroup stats?

Reply:
It doesn't, no, you explicitly have to go ask for your cgroup limits. Which makes some sense since you might still care about overall host capacity, and then the way you ask that would magically change inside a cgroup.

The stats part can depend on what's being read, too. You can read the same normal overall system usage, or you can read usage within a cgroup.

Reply:
Makes sense. I would have expected it to be a security issue to see things outside of your cgroup, but I see that's only with namespacing. If at all possible would like to fashion a test here. I wonder if we need to step back and think of a test harness that can test resource tuning in a CI way. (but approving anyways since this is still experimental)